### ISSUE TITLE
Running a .NET project using `AddProject` with the csproj path fails if the project has not been previously built

### ISSUE THREAD

**Issue by asimmon** [dotnet/aspire#2154]
I'm really glad that Preview 2 of Aspire added the `AddProject` overload so we can reference a .NET project by providing an arbitrary project path. However, I feel like Aspire currently assumes that this project is part of the app host solution, which would result in the project being built before being orchestrated by Aspire. If my project is not part of the app host solution and it has not been compiled yet, Aspire will fail to run it as it uses `--no-build` when preparing the projects.

In my opinion, this process somewhat defeats the purpose of using `AddProject` with a custom path because if my project is already part of my solution, I would simply use `AddProject<>` with the generated service metadata.

Is there any plan to consider building the .NET project prior to running it when using `AddProject` with a custom path? I believe this overload would be beneficial for scenarios where developers deal with multiple solutions in a monorepo (or even multiple repos).

**Comment by davidfowl** [comment]
Yes I think we should enable this. But you also want debugging inside of the IDE as well right?

**Comment by asimmon** [comment]
Yes, that would be great, although I'm not sure about the specifics of the experience. For now, building the project before running it would be enough.

When thinking about the debugging experience in .NET Aspire, I consider the possibility of instructing my app host code to attach the debugger to a specific resource right at its startup. On another note, I also think about a way to interact with the dashboard to manually attach to a process as needed. For now, I can already copy the process ID and manually attach to it from the IDE.

I remember that [Tye uses a host startup hook](https://github.com/dotnet/tye/blob/release/0.11.0/src/Microsoft.Tye.Hosting.Runtime/StartupHook.cs) to attach the debugger to the process at startup. But that's only valid for .NET apps.

**Comment by davidfowl** [comment]
Following up with some conversations the team has had about this topic. The solution we propose is building the project from the app host if it isn't part of the solution and adding support for attaching to arbitrary processes within VS. We think that those combination of fixes will address this gap.

**Comment by DamianEdwards** [comment]
> The solution we propose is building the project from the app host if it isn't part of the solution

Curious, how do we know that? Are we just assuming if the project is added via a .csproj path then it's not part of the solution and will try to build it, or do we have a more concrete signal? I'm concerned what will happen if we try to kick off a build of a project that's already being managed by VS.

**Comment by davidfowl** [comment]
I think that’s what we’d do. If it’s not added as a reference to the apphost then we build it. Not sure what else we can glean.

**Comment by rem-coding** [comment]
I know #1486 has been closed, as thought this was fixed for preview 5. But seems there's a potential regression in preview 6. Using the same example as previous with the AspireStarter template and changing to AddProject by path and removing appservice from the solution it fails as per the below screenshot:

<img width="1187" alt="image" src="https://github.com/dotnet/aspire/assets/39296150/7a9cbb9f-9c6c-44b5-9b91-6d066c93ded8">

**Comment by davidfowl** [comment]
This issue hasn't been fixed and won't be for the GA of Aspire. We've done some of the foundational work to make this possible but there are still some features that we need to build to make this possible.

**Comment by sherland** [comment]
As this is moved back to the backlog - Is there a workaround to this? Based on comments here and related issues, I get the impression that it worked at some time as long as the dependent project was first built manually?

So, is there any way to avoid the "Either the project type is not supported or the project is not loaded in the solution" error when starting debug from VS when the relevant project is not in the solution?

**Comment by davidfowl** [comment]
Moving this to 9 for investigation.

**Comment by davidfowl** [comment]
This won't be addressed for Aspire 9, moving to backlog. It's likely we will work on a bigger work item that will include this work for multi repo support.

**Comment by ChristophHornung** [comment]
I really wanted to use Aspire in our multi-repo environment so I added the ability to add external projects (just using dotnet run here) and made sure to wire up the same parts that `AddProject` does.

So if anybody wants to check this out before Aspire has official support for the scenario try the [nuget](https://www.nuget.org/packages/Chorn.Aspire.ExternalProject/1.0.0).

**Comment by davidfowl** [comment]
This is very cool @ChristophHornung !! We’ve discussed making this work in a similar way but changing the dcp protocol to make attach debugger work.

A few pieces of feedback (things we worried about using this technique):
- I see you’re using dotnet run, that’ll work well until you start running concurrently on projects in the same solution. AFAIK it’ll end up being uncoordinated and potentially may lead to conflicts/problems 
- The way you get the pid is problematic and will break in 9.1 (we redecorated all of this code so your private reflection will break) https://github.com/ChristophHornung/Aspire.ExternalProject/blob/9fd75e1d3e27ce878bbedec6d6d8dd0149471739/Chorn.Aspire.ExternalProject/ExternalProjectBuilderExtensions.cs#L147

For the first problem, one possible solution is to build up a single project file (msbuild file that references other projects) on the fly to compile multiple projects that may come from the same solution file and build that before starting the executable.

The season one you would use the resource notification service to get the pid instead of using private reflection. It should be available in a well know property (it shows in the dashboard)

**Comment by ChristophHornung** [comment]
@davidfowl Whoaw, thanks for the quick feedback.

On the topic of concurrent `dotnet run`s. I am not too versed on how the build works internally. I tried using the -o flag bug noticed that even when using `dotnet build -o ...` the obj folder was still created on referenced projects so I assumed it would not make much of a difference and just used `dotnet run`. I like the idea of stopping, git pull and starting again, this might be tricky when wrapping projects into a solution. I think I have to test that scenario a bit more.

Getting the pid is admittedly quick and dirty, and I saw that classes were renamed in the main branch. I did not see another way on first glance but will check out the resource notification service. Many thanks for pointing me in the right direction.

Could you elaborate on what you mean with `changing the dcp protocol to make attach debugger work`? This is the one area where I am really unhappy at the moment as the `vsjitdebugger -p <pid>` does not work correctly on any of my machines. (I can select my VS2022 instance but no breakpoints are hit). So I always have to manually attach from my VS instance searching through the list of processes.

**Comment by asimmon** [comment]
We built similar custom dotnet resource orchestration for external projects, to allow us going multi-repo. The main build concurrency issue we faced was when two or more dotnet processes tried to access a particular file for writing (like, a DLL). Our workaround so far was to use dotnet watch, build a custom lifecyclehook watching for build errors (`CS2012` for instance), and... bump the project file timestamp to force dotnet watch rebuild the project. It feels hacky, startup time may be increased, but it works.

We might explore the idea of generating a single parent csproj file at runtime to prevent this if this becomes too annoying for the devs.

**Comment by JimGeersinga** [comment]
How far are we into this feature?
Would love for this to work

**Comment by davidfowl** [comment]
See https://github.com/dotnet/aspire/issues/10920
